### PR TITLE
use oauth REST endpoints instead of raw etcd 

### PR DESCRIPTION
**Pull request by deads2k**
@mfojtik this is needed to remove direct etcd access which bypasses any admission checks and auditing and prevents the future separation of processes.

### CODE REVIEW

File: pkg/oauth/apiserver/auth.go [hunk 1]
```diff
@@ -156,13 +127,24 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
...
+	// glog.Infof("oauth server configured as: %#v", server)
+	// glog.Infof("auth handler: %#v", authHandler)
+	// glog.Infof("auth request handler: %#v", authRequestHandler)
+	// glog.Infof("grant checker: %#v", grantChecker)
+	// glog.Infof("grant handler: %#v", grantHandler)
+
+	return baseMux, nil
+}
+
+func (c *OAuthServerConfig) RegisterTokenRequestEndpoint(context genericapiserver.PostStartHookContext) error {
```
Comment (deads2k) [inline pkg/oauth/apiserver/auth.go]: This bit is nasty. The server won't report healthy until it succeeds though.  We need to decide where this should logically live, but I *think* this is the part where we actually depend on it.

File: pkg/cmd/server/origin/master.go [hunk 2]
```diff
@@ -180,20 +180,24 @@ func (c *MasterConfig) newAssetServerHandler() (http.Handler, error) {
...
 	}
 	oauthServer, err := config.Complete().New(apiserver.EmptyDelegate)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return oauthServer.GenericAPIServer.PrepareRun().GenericAPIServer.Handler.FullHandlerChain, nil
+	return oauthServer.GenericAPIServer.PrepareRun().GenericAPIServer.Handler.FullHandlerChain,
+		map[string]apiserver.PostStartHookFunc{
+			"oauth.openshift.io-RegisterTokenRequestEndpoint": config.RegisterTokenRequestEndpoint,
```
Comment (liggitt) [inline pkg/cmd/server/origin/master.go]: I wasn't expecting handler registration in post-start hooks... what requires this to be async?

File: pkg/cmd/server/origin/master_config.go [hunk 3]
```diff
@@ -696,7 +699,7 @@ func newAuthenticator(config configapi.MasterConfig, restOptionsGetter restoptio
 
 	// OAuth token
 	if config.OAuthConfig != nil {
-		oauthTokenAuthenticator, err := getEtcdTokenAuthenticator(restOptionsGetter, userGetter, groupMapper)
+		oauthTokenAuthenticator, err := getEtcdTokenAuthenticator(accessTokenGetter, userGetter, groupMapper)
```
Comment (liggitt) [inline pkg/cmd/server/origin/master_config.go]: should we be able to drop this at this point?

File: pkg/oauth/apiserver/auth.go (hunk 1 above)
Comment (liggitt) [inline pkg/oauth/apiserver/auth.go]: huh... adding a handler to the mux post-startup seems weirder than adding a handler and having it respond with errors until populated with the necessary clients post-startup

Comment (deads2k) [inline pkg/oauth/apiserver/auth.go]: > huh... adding a handler to the mux post-startup seems weirder than adding a handler and having it respond with errors until populated with the necessary clients post-startup

If you prefer.  How about I cut the links in this pull and twiddle it after.  This is the big money and rest is arguing over spoils.

File: pkg/cmd/server/origin/master.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: > I wasn't expecting handler registration in post-start hooks... what requires this to be async?

The creation of clients needs to be a post-start hook, the handler could give failures until they exist.

File: pkg/cmd/server/origin/master_config.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > should we be able to drop this at this point?

I wasn't anticipating dropping anything.  I'll look back and see.

**Comment by deads2k** [comment]
Comments addressed

File: pkg/auth/server/grant/grant_test.go [hunk 4]
```diff
@@ -402,25 +420,37 @@ func TestGrant(t *testing.T) {
...
-			auth := testCase.AuthRegistry.UpdatedAuthorization
-			if auth == nil {
+			found := false
+			for _, action := range testCase.AuthRegistry.Actions() {
+				if action.Matches("update", "oauthclientauthorizations") {
+					found = true
+					auth := action.(clienttesting.UpdateAction).GetObject().(*oapi.OAuthClientAuthorization)
+					if !reflect.DeepEqual(testCase.ExpectUpdatedAuthScopes, auth.Scopes) {
+						t.Errorf("%s: expected updated scopes %v, got %v", k, testCase.ExpectUpdatedAuthScopes, auth.Scopes)
+					}
```
Comment (mfojtik) [inline pkg/auth/server/grant/grant_test.go]: bit: break ?

File: pkg/auth/server/grant/grant_test.go [hunk 5]
```diff
@@ -402,25 +420,37 @@ func TestGrant(t *testing.T) {
...
-			auth := testCase.AuthRegistry.CreatedAuthorization
-			if auth == nil {
+			found := false
+			for _, action := range testCase.AuthRegistry.Actions() {
+				if action.Matches("create", "oauthclientauthorizations") {
+					found = true
+					auth := action.(clienttesting.CreateAction).GetObject().(*oapi.OAuthClientAuthorization)
+					if !reflect.DeepEqual(testCase.ExpectCreatedAuthScopes, auth.Scopes) {
+						t.Errorf("%s: expected created scopes %v, got %v", k, testCase.ExpectCreatedAuthScopes, auth.Scopes)
+					}
```
Comment (mfojtik) [inline pkg/auth/server/grant/grant_test.go]: nit: break?

File: pkg/cmd/server/origin/master.go [hunk 6]
```diff
@@ -180,20 +180,24 @@ func (c *MasterConfig) newAssetServerHandler() (http.Handler, error) {
...
 	}
 	oauthServer, err := config.Complete().New(apiserver.EmptyDelegate)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return oauthServer.GenericAPIServer.PrepareRun().GenericAPIServer.Handler.FullHandlerChain, nil
+	return oauthServer.GenericAPIServer.PrepareRun().GenericAPIServer.Handler.FullHandlerChain,
+		map[string]apiserver.PostStartHookFunc{
+			"oauth.openshift.io-EnsureBootstrapOAuthClients": config.EnsureBootstrapOAuthClients,
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master.go]: shouldn't this be a const?

File: pkg/oauth/apiserver/auth.go [hunk 7]
```diff
@@ -157,21 +126,14 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
...
-	osOAuthClientConfig := c.NewOpenShiftOAuthClientConfig(browserClient)
+	osOAuthClientConfig := newOpenShiftOAuthClientConfig(OpenShiftBrowserClientID, "", c.Options.MasterPublicURL, c.Options.MasterURL)
 	osOAuthClientConfig.RedirectUrl = c.Options.MasterPublicURL + path.Join(oauthutil.OpenShiftOAuthAPIPrefix, tokenrequest.DisplayTokenEndpoint)
 
 	osOAuthClient, _ := osincli.NewClient(osOAuthClientConfig)
 	if len(*c.Options.MasterCA) > 0 {
 		rootCAs, err := cmdutil.CertPoolFromFile(*c.Options.MasterCA)
 		if err != nil {
-			glog.Fatal(err)
+			return nil, err
```
Comment (mfojtik) [inline pkg/oauth/apiserver/auth.go]: @deads2k i hope this error is properly captured in upper level :)

File: pkg/oauth/apiserver/oauth_apiserver.go [hunk 8]
```diff
@@ -185,3 +194,42 @@ func (c *OAuthServerConfig) buildHandlerChainForOAuth(startingHandler http.Handl
 	handler = genericfilters.WithPanicRecovery(handler)
 	return handler
 }
+
+// TODO, this moves to the `apiserver.go` when we have it for this group
+// TODO TODO, this actually looks a lot like a controller or an add-on manager style thing.  Seems like we'd want to do this outside
```
Comment (mfojtik) [inline pkg/oauth/apiserver/oauth_apiserver.go]: metatodo? :)

Comment (mfojtik) [reply pkg/oauth/apiserver/oauth_apiserver.go]: also lacks normal godoc

File: pkg/oauth/registry/oauthclientauthorization/strategy.go [hunk 9]
```diff
@@ -137,3 +137,7 @@ func Matcher(label labels.Selector, field fields.Selector) kstorage.SelectionPre
 func SelectableFields(obj *oauthapi.OAuthClientAuthorization) fields.Set {
 	return oauthapi.OAuthClientAuthorizationToSelectableFields(obj)
 }
+
+func ClientAuthorizationName(userName, clientName string) string {
```
Comment (mfojtik) [inline pkg/oauth/registry/oauthclientauthorization/strategy.go]: godoc?

File: test/integration/master_routes_test.go [hunk 10]
```diff
@@ -99,6 +99,7 @@ var expectedIndex = []string{
 	// "/healthz/poststarthook/extensions/third-party-resources",  // Do not enable this controller, we do not support it
 	"/healthz/poststarthook/generic-apiserver-start-informers",
 	"/healthz/poststarthook/kube-apiserver-autoregistration",
+	"/healthz/poststarthook/oauth.openshift.io-EnsureBootstrapOAuthClients",
```
Comment (mfojtik) [inline test/integration/master_routes_test.go]: why is this camelcase while the rest is lowecase?

File: test/integration/oauthstorage_test.go [hunk 11]
```diff
@@ -58,31 +52,22 @@ func TestOAuthStorage(t *testing.T) {
...
-	optsGetter, err := originrest.StorageOptions(*masterOptions)
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMasterAPI(masterOptions)
 	if err != nil {
-		t.Fatalf("unexpected error: %v", err)
-	}
-
-	clientStorage, err := clientetcd.NewREST(optsGetter)
-	if err != nil {
-		t.Fatalf("unexpected error: %v", err)
+		t.Fatal(err)
```
Comment (mfojtik) [inline test/integration/oauthstorage_test.go]: now we won't know this was unexpected :-)

**Comment by mfojtik** [comment]
just couple small nits, but the intent and how this was done LGTM

/approve

File: test/integration/master_routes_test.go (hunk 10 above)
Comment (deads2k) [reply test/integration/master_routes_test.go]: > why is this camelcase while the rest is lowecase?

we're working through what it should look like.

File: pkg/oauth/apiserver/auth.go (hunk 7 above)
Comment (deads2k) [reply pkg/oauth/apiserver/auth.go]: > @deads2k i hope this error is properly captured in upper level :)

it is

File: pkg/auth/server/grant/grant_test.go (hunk 4 above)
Comment (deads2k) [reply pkg/auth/server/grant/grant_test.go]: done

File: pkg/auth/server/grant/grant_test.go (hunk 5 above)
Comment (deads2k) [reply pkg/auth/server/grant/grant_test.go]: done

File: pkg/oauth/apiserver/oauth_apiserver.go (hunk 8 above)
Comment (deads2k) [reply pkg/oauth/apiserver/oauth_apiserver.go]: done

File: pkg/oauth/registry/oauthclientauthorization/strategy.go (hunk 9 above)
Comment (deads2k) [reply pkg/oauth/registry/oauthclientauthorization/strategy.go]: done

**Comment by deads2k** [comment]
comments addressed. applying label.

**Pull request merged**